File: Sep/Day18/design-task-manager.py
import heapq
from collections import defaultdict
from typing import List

class TaskManager:
    '''
    TC : add: O(log N),edit: O(log N) ,rmv: O(1) (no more scanning over all users!)
    execTop: amortized O(log N)
    SC - o(N)
    '''
    def __init__(self, tasks: List[List[int]]):
        self.taskMap = {}                # taskId -> priority
        self.globalHeap = []             # heap of (-priority, taskId)

        for userId, taskId, priority in tasks:
            self.add(userId, taskId, priority)

    def add(self, userId: int, taskId: int, priority: int) -> None:
        self.taskMap[taskId] = (priority, userId)
        heapq.heappush(self.globalHeap, (-priority, -taskId))

    def edit(self, taskId: int, newPriority: int) -> None:
        if taskId in self.taskMap:
            _, userId  = self.taskMap[taskId]
            self.taskMap[taskId] = (newPriority, userId)
            heapq.heappush(self.globalHeap, (-newPriority, -taskId))

    def rmv(self, taskId: int) -> None:
        _, userId = self.taskMap[taskId]
        del self.taskMap[taskId]

    def execTop(self) -> int:
        while self.globalHeap:
            priority, taskId = heapq.heappop(self.globalHeap)
            priority = -priority
            taskId = -taskId
            if taskId in self.taskMap and self.taskMap[taskId][0] == priority:
                _, userId = self.taskMap[taskId]
                self.rmv(taskId)
                return userId
        return -1  # no tasks left

# Your TaskManager object will be instantiated and called as such:
# obj = TaskManager(tasks)
# obj.add(userId,taskId,priority)
# obj.edit(taskId,newPriority)
# obj.rmv(taskId)
# param_4 = obj.execTop()

File: Sep/Day18/design-task-manager.java
import java.util.*;

class TaskManager {
    private Map<Integer, int[]> taskMap; // taskId -> [priority, userId]
    private PriorityQueue<int[]> globalHeap;

    public TaskManager(List<int[]> tasks) {
        taskMap = new HashMap<>();
        globalHeap = new PriorityQueue<>((a, b) -> {
            if (b[0] != a[0]) return b[0] - a[0]; // higher priority first
            return b[1] - a[1]; // higher taskId first
        });

        for (int[] t : tasks) {
            add(t[0], t[1], t[2]);
        }
    }

    public void add(int userId, int taskId, int priority) {
        taskMap.put(taskId, new int[]{priority, userId});
        globalHeap.offer(new int[]{priority, taskId});
    }

    public void edit(int taskId, int newPriority) {
        if (taskMap.containsKey(taskId)) {
            int userId = taskMap.get(taskId)[1];
            taskMap.put(taskId, new int[]{newPriority, userId});
            globalHeap.offer(new int[]{newPriority, taskId});
        }
    }

    public void rmv(int taskId) {
        taskMap.remove(taskId);
    }

    public int execTop() {
        while (!globalHeap.isEmpty()) {
            int[] top = globalHeap.poll();
            int priority = top[0], taskId = top[1];
            if (taskMap.containsKey(taskId) && taskMap.get(taskId)[0] == priority) {
                int userId = taskMap.get(taskId)[1];
                rmv(taskId);
                return userId;
            }
        }
        return -1;
    }
}
